### ISSUE TITLE
Bug: Type error when using ConfigWithExtends

### ISSUE THREAD

**Issue by doberkofler** [typescript-eslint/typescript-eslint#8732]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-eslint

### Repro Code

I get a TypeScript type error `scripts-eslint/eslint.config.mjs:7:2 - error TS2353: Object literal may only specify known properties, and 'ignores' does not exist in type 'ConfigWithExtends'` when using the `ConfigWithExtends` type in the following example. 

```TypeScript
// @ts-check

import tseslint from 'typescript-eslint';

/** @type {import('typescript-eslint').ConfigWithExtends} */
const config_base = {
	ignores: [
		'temp/**/*',
	],
};

/** @type {import('typescript-eslint').ConfigWithExtends[]} */
const configs = [config_base];

const flat_config = tseslint.config(...configs);

export {flat_config};
```

### Expected Result

n.a.

### Actual Result

n.a.

### Additional Info

[flatconfig_error_1.zip](https://github.com/typescript-eslint/typescript-eslint/files/14655262/flatconfig_error_1.zip)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`             | `7.3.1` |
| `TypeScript`                           | `5.4.3` |
| `ESLint`                               | `8.57.0` |
| `node`                                 | `20.11.0` |

**Comment by peachest** [comment]
I met the same question. I was learning to use the eslint flat config and suddenly the Webstorm told me that the `TSESLint.FlatConfig` type is not exported. Some time before this happened, I had checked this declaration file and at that time everything is alright and I can jump to check the FlatConfig type declaration.
See the below screenshot.
![image](https://github.com/typescript-eslint/typescript-eslint/assets/75366539/a8e36a6d-fa17-4d2c-b089-0bb272564925)
And then I checked the history of file `@typescript-eslint/utils/dist/ts-eslint/index.d.ts` file, it seems like at one time, this file changed, but I am sure it's not by me since the Reader Mode for node_module file is still on.
<img width="1099" alt="image" src="https://github.com/typescript-eslint/typescript-eslint/assets/75366539/8cf5682f-746e-416d-ab48-3af811a467c6">
The export of `./Config` is deleted and for sure the type is not alright.
I think this may caused by the package update or something like this.

**Comment by peachest** [comment]
> ### Before You File a Bug Report Please Confirm You Have Done The Following...
> 在提交错误报告之前，请确认您已完成以下操作...
> * [x]  I have tried restarting my IDE and the issue persists. 我尝试重新启动 IDE，但问题仍然存在。
> * [x]  I have updated to the latest version of the packages. 我已经更新到最新版本的软件包。
> * [x]  I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue. 我搜索了相关问题，但没有发现与我的问题相符的问题。
> * [x]  I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed. 我已阅读常见问题解答，但我的问题未列出。
> 
> ### Relevant Package 相关套餐
> typescript-eslint
> 
> ### Playground Link 游乐场链接
> _No response 没有反应_
> 
> ### Repro Code 复制代码
> ```ts
> I get a TypeScript type error `scripts-eslint/eslint.config.mjs:7:2 - error TS2353: Object literal may only specify known properties, and 'ignores' does not exist in type 'ConfigWithExtends'` when using the `ConfigWithExtends` type in the following example.
> 
> 
> // @ts-check
> 
> import tseslint from 'typescript-eslint';
> 
> /** @type {import('typescript-eslint').ConfigWithExtends} */
> const config_base = {
> 	ignores: [
> 		'temp/**/*',
> 	],
> };
> 
> /** @type {import('typescript-eslint').ConfigWithExtends[]} */
> const configs = [config_base];
> 
> const flat_config = tseslint.config(...configs);
> 
> export {flat_config};
> ```
> 
> ```
> 
> 
> ### ESLint Config
> 
> _No response_
> 
> ### tsconfig
> 
> _No response_
> 
> ### Expected Result
> 
> n.a.
> 
> ### Actual Result
> 
> n.a.
> 
> ### Additional Info
> 
> [flatconfig_error_1.zip](https://github.com/typescript-eslint/typescript-eslint/files/14655262/flatconfig_error_1.zip)
> 
> 
> ### Versions
> 
> 
> | package                                | version |
> | -------------------------------------- | ------- |
> | `typescript-eslint`             | `7.3.1` |
> | `TypeScript`                           | `5.4.3` |
> | `ESLint`                               | `8.57.0` |
> | `node`                                 | `20.11.0` |
> ```

Now the typescript can only see the `extends` field in `ConfigWithExtends` interface，this explains why other field of `Config` including `ignores` will be treated as no exist.

**Comment by doberkofler** [comment]
I just discovered that omitting the `eslint-plugin-jest` dependency changes the behaviour (no type error) but I have no idea why.

**Comment by bradzacher** [comment]
@peachest your workspace is in a bad state (the `@typescript-eslint/utils` version is out of date) and is thus unrelated to this.

**Comment by bradzacher** [comment]
In the  example you've linked @doberkofler for some reason `@typescript-eslint/utils` v5 has been installed. It looks like this is just a HORRENDOUS module installation from `npm`.

`eslint-plugin-jest` depends on v5 of our tooling.
`typescript-eslint` depends on v7 of our tooling.

For some reason `npm` is seeing the v5 dependency and only installing that version.
It's completely disrespecting the declared dependency versions.

Try regenerating your lockfile from scratch to force `npm` to actually respect things.
I just deleted it and re-installed and it did what it was supposed to - installed v7 at the root and v5 as a child of `eslint-plugin-jest`

**Comment by doberkofler** [comment]
I'm sure your findings are correct but I'm not able to resolve this problem.
Already when creating this reduced example I each time removed the `package-lock.json` file and the node_modules directory to guarantee a clean installation.
I just again to make sure and the error persists.
I use node 20.11.0 and npm 10.5.0.

**Comment by doberkofler** [comment]
@bradzacher This is the generated lock file: [package-lock.json.zip](https://github.com/typescript-eslint/typescript-eslint/files/14657699/package-lock.json.zip)

**Comment by bradzacher** [comment]
If you check the version listed in `./node_modules/@typescript-eslint/utils/package.json` it should be v7

If it's v5 then double check `./node_modules/typescript-eslint` - it should contain its own `./node_modules/@typescript-eslint/utils` which is v7.

If you don't have a v7 version of the utils package - then that's the cause - npm is doing the wrong thing.

**Comment by doberkofler** [comment]
> If you check the version listed in `./node_modules/@typescript-eslint/utils/package.json` it should be v7

It seems to be version  5

> If it's v5 then double check `./node_modules/typescript-eslint` - it should contain its own `./node_modules/@typescript-eslint/utils` which is v7.

It does not seem to contain a `./node_modules/@typescript-eslint/utils` directory

> If you don't have a v7 version of the utils package - then that's the cause - npm is doing the wrong thing.

I understand:
Would you know if this is a known npm problem?
Does the `eslint-plugin-jest` need to update to v7?
Is there any other workaround?

**Comment by peachest** [comment]
My simple reproduce
1. I delete both dependencies, yarn.lock and node_module for reinstalling all package, then the utils/ is correct
2. I add eslint-plugin-jest. In the lock file eslint-plugin-jest depent the v5 utils/, but strangely the v7 type declaration is also available and IDE will jump to the v7 utils/. I have try close my IDE, delete the .idea/ and restart the IDE to force re-indexing the project but still the IDE find the  v7 utils/. So strange!
3. I add eslint-plugin-jest-extended, finally the IDE now find the v5 utils/
![image](https://github.com/typescript-eslint/typescript-eslint/assets/75366539/80666902-32e3-463f-ae80-8538111517a8)

I can provide an ugly solution for this, just force typescript use v7 @typescript-eslint/utils using `resolutions` in your package.json. This work for me
```json
  "devDependencies": {
    "@eslint/js": "^8.57.0",
    "eslint": "^8.57.0",
    "eslint-plugin-jest": "^27.9.0",
    "eslint-plugin-jest-extended": "^2.0.0",
    "typescript": "^5.4.2",
    "typescript-eslint": "^7.3.1"
  },
  "resolutions": {
    "@typescript-eslint/utils": "^7.0.0"
  }
```

**Comment by bradzacher** [comment]
> Would you know if this is a known npm problem?

I honestly don't know, sorry. I don't use npm myself (I bounce around between yarn1, yarn2 and pnpm) so I don't know if this is expected or not.
It seems like very, very, very wrong behaviour.

> Does the eslint-plugin-jest need to update to v7?

Yes they should! They're working on it - https://github.com/jest-community/eslint-plugin-jest/pull/1508
They have some difficulties because they support old versions of node and old versions of eslint (which we do not support).

> Is there any other workaround?

If you're using npm 9 - https://docs.npmjs.com/cli/v9/configuring-npm/package-json#overrides
Yarn/pnpm supports `resolutions` as @peachest notes.
You could also try just adding `@typescript-eslint/utils: "^7"` to your devDeps - I'd assume this will force npm to respect that you really really want v7.

**Comment by doberkofler** [comment]
I've tried adding the resolution but unfortunately it does not solve the problem for me.
I'm using the latest npm version 10.5.0.
I filed a npm SR referring to this SR.

**Comment by doberkofler** [comment]
I have also tried using:
```
	"overrides": {
		"typescript-eslint": {
			"@typescript-eslint/utils": "^7.0.0"
		}
	}
```
but it also does not work.

**Comment by nlwillia** [comment]
@bradzacher when I `npm i -D typescript-eslint` into a clean project in npm 10.1.0, I get `node_modules/@typescript-eslint/utils` where the package version is 7.3.1.  But when I look at the `dist\ts-eslint\Config.d.ts` in that package, the FlatConfig.Config interface is not exported even though it is exported in the [main source](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/utils/src/ts-eslint/Config.ts#L247).  I don't understand why that would be when the ClassicConfig.Config interface earlier in the file is exported correctly.  I tried building from source myself, and it does the same.

**Comment by bradzacher** [comment]
The more I investigate the more I'm **CONVINCED** that npm is doing the wrong thing here.
But that's ultimately neither here nor there because that's just how it works.

#8738 should fix this by explicitly declaring the dep so that npm at least nests `/utils` within `typescript-eslint` so types will work.

**Comment by AndreaPontrandolfo** [comment]
@bradzacher i have a pnpm workspace and i'm encountering the same problem. this is definitely unexpected behaviour.
